### ISSUE TITLE
Discrepency in the number of generations used between official "develop" branch and Scott LeGrand's "relicensing" branch

### ISSUE THREAD

**Issue by RemiLacroix-IDRIS** [ccsb-scripps/AutoDock-GPU#101]
Hi,

I am running some tests with AutoDock-GPU, trying to reproduce some benchmark results published in https://arxiv.org/pdf/2007.03678.pdf (section 3). Input data can be found at https://github.com/diogomart/AD-GPU_set_of_42.

I get very different performances (i.e. total time to process for the whole set) when using the [official "develop" branch](https://github.com/ccsb-scripps/AutoDock-GPU/tree/develop) and [Scott LeGrand's "relicensing" branch](https://github.com/scottlegrand/AutoDock-GPU/tree/relicensing):
- official "develop" branch : 698.193 s without overlapping, 556.813 s with overlapping
- @scottlegrand's "relicensing" branch : 354.593 s without overlapping, 219.714 s with overlapping.

While trying to track down the cause of this, I noticed in the "dlg" output files that the number of generations used was different for both versions.

Here is a example for one of the protein-ligand pairs from the test set:
- official "develop" branch:
> Number of energy evaluations performed:    2512598
> Number of generations used:                113
- Scott LeGrand's "relicensing" branch:
> Number of energy evaluations performed:    2530580
> Number of generations used:                70

Any idea to explain this behavior?

Best,
Rémi

**Comment by scottlegrand** [comment]
What happens if you run the same search on the develop branch twice? Do you
get the same # of evaluations and generations?  If there's any sort of
convergence testing that cuts off prematurely you should not because that
code is not deterministic.

Humor me?

Scott

**Comment by scottlegrand** [comment]
But even then, CUDA and OpenCL are not the same codepath and that will make
things different if the number of evaluations and generations are anything
but 100% hardcoded static parameters just from arbitrarily different orders
of seemingly innocuous operations. See also different RNG initialization.

But even if you account for that, the associative floating point roundoff
error of FP32 atomic adds will cause differences between runs unless those
parameters are 100% static just because chaos is as chaos does. AI people
see this all the time with the major deep learning frames as an aside.

They even use that as a substitute for multiple samples, which it is not
because while there is chaos, if the initializations are identical, there's
a huge amount of hard to rigorously describe correlation.

But first things first, I need more data here if you need to go down this
rabbit hole.

**Comment by RemiLacroix-IDRIS** [comment]
> But even then, CUDA and OpenCL are not the same codepath

Note that I am comparing your CUDA version to the CUDA version that have been merged in Scripps' repo.

> What happens if you run the same search on the develop branch twice? Do you get the same # of evaluations and generations?

It seems the numbers of evaluations and generations are always the same in your branch. There are some variations when using Scripps' branch but the numbers are always the same magnitude.

**Comment by scottlegrand** [comment]
That makes sense, my code is 100% deterministic given the same initial
conditions. If what has been merged is showing variance, then something
didn't get merged... For example, if you are initializing the RNG from the
clock rather than using the same seed, that will cause nondeterminism.

If you change the seed in my branch, you should either see variance, or the
convergence tests between my branch and the main are different.

**Comment by scottlegrand** [comment]
Second, there's not enough information here to say there's anything wrong
either way. They are doing about the same number of energy evaluations, and
I gather generating pretty much the same quality results. Beyond any
convergence test differences, the dynamic range handling of fixed point
accumulation could conceivably create longer searches that lead to fewer
generations needed. Again, not asserting anything here whatsoever. But
unless there's a qualitative difference in the results, I don't know how to
proceed further with this beyond looking at the convergence tests. I do
know that fixed point force accumulation creates better energy conservation
than FP32 accumulation, but I am in no way asserting something similar is
going on here, just noting that there literally might be nothing to see
here.

**Comment by RemiLacroix-IDRIS** [comment]
> Second, there's not enough information here to say there's anything wrong either way. They are doing about the same number of energy evaluations, and I gather generating pretty much the same quality results.

Well, I am mostly concerned by the performance difference between Scripps' version and yours.

> I get very different performances (i.e. total time to process for the whole set) when using the official "develop" branch and Scott LeGrand's "relicensing" branch:
> - official "develop" branch : **698.193 s** without overlapping, 556.813 s with overlapping
> - @scottlegrand's "relicensing" branch : **354.593 s** without overlapping, 219.714 s with overlapping.

They are doing about the same number of energy evaluations and are basically 2 times slower...

I probably should have phrased this issue differently. My question is: "why is Scripps' version slower and can the different generation numbers be a good explanation for the performance difference?"

**Comment by scottlegrand** [comment]
So this is the speed difference between the merged version of my code and
my branch? Oh boy... Yep, there's a problem.

**Comment by atillack** [comment]
@scottlegrand Are you able to verify the runtime difference on one of your machines?

@RemiLacroix-IDRIS I take it you are running with everything being at default values, right? In other words, Solis-Wets (sw) and 2.5 M evals.

I have just run Scott's current relicensing branch and our current branch on the same node with a Titan V and cannot see any dramatic performance difference for Diogo's 42 with SW and 2.5 M evals. See for yourself:

Scott's current relicensing branch (Cuda):
# stats:        42                       105000000       23  /  42 good                           560.35 s      97.65 s

Our current develop branch (Cuda):
# stats:        42                       105000000       31  /  42 good                           544.24 s      97.22 s

All of these were run back-to-back on the same machine with no background processes. The first number is the number of ligands (this is Diogo's 42, so no surprise there), the overall number of evals (2.5 M * 42 = 105 M), how many energies are within acceptable ranges (our version has a slightly updated SW algorithm, so no surprises there), overall compute time, and setup times.

Please let me know what hardware, software, and command line options you are running as it would be nice to finally figure out why Cuda's SW is slower by a factor of two compared to OpenCL typically. Oh boy, indeed...

**Comment by atillack** [comment]
@RemiLacroix-IDRIS @scottlegrand If you are running with Adadelta (ad) then the discrepancy is likely that Scott hasn't taken out what we called the "restoring gradient". It will make a run finish quicker but it comes with the huge downside of equally worse statistics which is why we disabled it by default. In docking, getting worse results quicker is usually not what we want ...

**Comment by atillack** [comment]
@RemiLacroix-IDRIS @scottlegrand Just for completeness' sake, this is the OpenCL run on the exact same machine with the same settings (sw, 2.5 M evals):
# stats:        42                       105000000       33  /  42 good                           220.02 s      95.95 s

**Comment by atillack** [comment]
@scottlegrand If I remember correctly, the 2.9x speedup that you observed is for Adadelta with the restoring gradient. The 1.5x I am seeing is without. Since I wrote the restoring gradient it took some pretty good evidence to convince me to be fine disabling it ... Maybe @diogomart can chime in when he has a moment.

**Comment by atillack** [comment]
@scottlegrand Because they can't afford you?

**Comment by atillack** [comment]
@scottlegrand Anyway, jokes aside. Unfortunately, I can't reproduce the behavior at this point as a little bit more Information about the runs is needed. If there's a performance regression we'll fix it.

**Comment by scottlegrand** [comment]
No, they pay the people that blacklisted me far more. But the next time
their search engine gives you useless dystopic results, think about them,
not me.

**Comment by diogomart** [comment]
The restoring gradient had a very large impact on the quality of the results (convergence). I'm gonna do the convergence tests and come back with data.

**Comment by diogomart** [comment]
@RemiLacroix-IDRIS @scottlegrand @atillack please confirm these are the commits of interest:

bdb6b22 from Scott's relicensing branch
9a0d852 from Scripps' develop branch

**Comment by atillack** [comment]
@diogomart I can confirm that those are the ones I ran for the runs above.

**Comment by diogomart** [comment]
Here's data for 36 out of 42 systems in the set. The 6 excluded ligands are larger. Each system was docked 4 times and 100 GA runs with a fixed number of evaluations, from 32k, 64k, 128k to 8192k (9 data points). The plot below reports the average compute time (excluding idle time) out of the 4 dockings. There's a total of 36*9 data points. The differences in compute time between bdb6b22 and 9a0d852 are very small.

The second plot are the convergence tests. The values in the axis are the number of evaluations (in thousands) needed for a single GA run to achieve a 50% success rate, which is to be within 1 kcal/mol from the best solution, or within 2 angstrom RMSD from the x-ray pose. @scottlegrand's relicensing branch (bdb6b22) requires more evaluations to achieve the 50% success rate.

![9a0d852-cuda-128wi-ad--bdb6b22-cuda-128wi-ad runtime](https://user-images.githubusercontent.com/1564797/90968447-2cc1fe00-e4a1-11ea-907b-2d64e46f6e3b.png)

![9a0d852-cuda-128wi-ad_vs_bdb6b22-cuda-128wi-ad](https://user-images.githubusercontent.com/1564797/90968494-8cb8a480-e4a1-11ea-91fc-4592e92479e0.png)

**Comment by diogomart** [comment]
This is using adadelta. The GPU is a titan V

**Comment by atillack** [comment]
@diogomart Thank you!

@scottlegrand yes. we concluded that the restoring gradient essentially keeps runs alive that will likely fail again, and again, and again - sometimes cheaper but less useful.

**Comment by RemiLacroix-IDRIS** [comment]
> If you are running with Adadelta (ad) then the discrepancy is likely that Scott hasn't taken out what we called the "restoring gradient". It will make a run finish quicker but it comes with the huge downside of equally worse statistics which is why we disabled it by default. In docking, getting worse results quicker is usually not what we want ...

Could be, I was indeed using Adadelta.

**Comment by diogomart** [comment]
@RemiLacroix-IDRIS what GPU and work size are you using? Also, could you post all the command line parameters, e.g. -nev, -nrun, -lsit, -lsrat, and so on.

@scottlegrand I'll report run times for numwi=64 soon

**Comment by RemiLacroix-IDRIS** [comment]
> @RemiLacroix-IDRIS what GPU and work size are you using?

NVIDIA V100 with numwi=64.

> Also, could you post all the command line parameters, e.g. -nev, -nrun, -lsit, -lsrat, and so on.

Just `autodock_gpu_64wi -filelist batchfile -nrun 10 -lsmet ad` so mostly default settings.

**Comment by diogomart** [comment]
@scottlegrand 64 work items decreases compute time by ~ 25 % for your branch, and by ~ 15% for our branch. 

@RemiLacroix-IDRIS the difference in the number of generations is due to a stopping criterion for adadelta. It is used by default in our branch `#define ADADELTA_AUTOSTOP`, but not in @scottlegrand's branch `// #define AD_RHO_CRITERION`. For ligands with few rotatable bonds, say less than 5, we observed convergence in about half the number of evaluations with the stopping criterion, that's why we turned it on. The increase in runtime that you observe only occurs for small values of `-nrun`, such as 10. In the plots I posted above, `-nrun` was set to 100, hence no meaningful differences in runtime with or without the adadelta stopping criterion.

**Comment by atillack** [comment]
@diogomart @RemiLacroix-IDRIS @scottlegrand Just to add to what Diogo was writing - when I enable the restoring gradient in our branch for Adadelta I get about the same overall runtime compared to Scott's branch.

We will likely revisit this idea in the not too distant future but as you've seen right now, we chose a better statistical chance at getting well converged results over quicker but potentially worse results. Also, as you may have seen already, we are working on integrating Scott's determinism code which will also yield a nice performance boost. As with the previous integrations, we will make sure that we don't regress in terms of convergence.

Thank you Remi for your eagle-eyed observations and for posting what you find. Also, thank you everybody for the fruitful discussion :-)

**Comment by RemiLacroix-IDRIS** [comment]
@atillack Can you confirm that I can enable restoring grading by uncommenting `// #define RESTORING_MAP_GRADIENT`? I would like to check whether the impact of restoring grading is the same on the CUDA version and OpenCL version.

**Comment by diogomart** [comment]
@RemiLacroix-IDRIS keep in mind that it will probably impact runtime only if `ADADELTA_AUTOSTOP` (or `ADADELTA_RHO_CRITERION` in Scott's branch) are also defined.

**Comment by RemiLacroix-IDRIS** [comment]
Here are the comparisons between the default configuration and all combinations of autostop (edit: `ADADELTA_AUTOSTOP` not `-autostop`) and restoring gradient for both the CUDA and OpenCL versions.

Edit: All settings are default expect for nrun which is set to 10.

The values in parentheses are the accelerations compared to the default configuration.
 
CUDA version with                        | No overlap  | Overlap (10 threads)
-----------------------------------------|:-----------:|:----------:
Default settings<br>autostop: YES<br>restoring gradient: NO  | 698.193 sec (1.00x) | 556.813 sec (1.25x)
autostop: YES<br>restoring gradient: YES | 470.707 sec (1.48x) | 339.024 sec (2.06x)
autostop: NO<br>restoring gradient: YES  | 360.622 sec (1.94x) | 223.035 sec (3.13x)
autostop: NO<br>restoring gradient: NO   | 365.152 sec (1.91x) | 226.563 sec (3.08x)

OpenCL version with                      | No overlap  | Overlap (10 threads)
-----------------------------------------|:-----------:|:----------:
Default settings<br>autostop: YES<br>restoring gradient: NO  | 969.323 sec (1.00x) | 836.040 sec (1.16x)
autostop: YES<br>restoring gradient: YES | 973.001 sec (1.00x) | 845.206 sec (1.15x)
autostop: NO<br>restoring gradient: YES  | 603.656 sec (1.61x) | 479.573 sec (2.02x)
autostop: NO<br>restoring gradient: NO   | 636.629 sec (1.52x) | 480.623 sec (2.02x)

It seems the fastest combination is no autostop + restoring gradient and as far as I remember that's the default for Scott's branch.

If you compare the CUDA and OpenCL version for each combinations, you get the following accelerations:

Configuration                            | No overlap  | Overlap (10 threads)
-----------------------------------------|:-----------:|:----------:
Default settings<br>autostop: YES<br>restoring gradient: NO  |       1.39x |       1.50x
autostop: YES<br>restoring gradient: YES |       2.07x |       2.49x
autostop: NO<br>restoring gradient: YES  |       1.67x |       2.15x
autostop: NO<br>restoring gradient: NO   |       1.74x |       2.12x

It shows that the impact of having autostop and/or restoring gradiend enabled is different for the CUDA and OpenCL versions.

/cc @asedova

**Comment by diogomart** [comment]
@RemiLacroix-IDRIS thank you for posting this comprehensive and well organized data. This is very informative.

Just to be 100% sure, `autostop` refers to `ADADELTA_AUTOSTOP` and not the command line option `-autostop`, correct?

@atillack at this very low `-nrun` regime (default=1) it seems that the improved convergence that we observed with `ADADELTA_AUTOSTOP` (about 2x for easier systems) mostly cancels out with the slower runtime.

**Comment by RemiLacroix-IDRIS** [comment]
> Just to be 100% sure, `autostop` refers to `ADADELTA_AUTOSTOP` and not the command line option `-autostop`, correct?

Yes, it does refer to `ADADELTA_AUTOSTOP`, sorry about that, I forgot about the `-autostop` option.

> @atillack at this very low `-nrun` regime (default=1)

I should have added that the results above are for `-nrun 10`.

**Comment by atillack** [comment]
@RemiLacroix-IDRIS I do see similar relative behavior. One thing that puzzles me a little is your V100 is about 1.5x slower compared to our Titan V: When I run with your settings with NUMWI=64, OVERLAP, without Restoring Gradient, and with ADADELTA_AUTOSTOP in Cuda, it takes our Titan V 233.64 s + 3.17 s = 236.81 s compared to your 339.024 s. The runs used Cuda 10.2 and GCC 6.3.1. While this may just be the difference between active and passive cooling, it nonetheless is quite a strong difference for some very similar cards.

When looking at the runtime of these options, convergence needs to be taken into account as well. This is why we set those settings (restoring gradient and Adadelta autostop) in the way they are. Particularly for lower run numbers convergence will make or break a virtual screening (usually of compounds you don't have reference data at hand for). The runtime differences observed when changing those settings are due to worse search performance and would vanish if more evaluations to equal overall search performance were performed.

To put it differently, if this were a race, without knowing how far a participant went you couldn't conclude that they went twice as fast only because they stopped after half the time. This is the case with these options, changing them usually also means only going half the distance or worse on average all while expending the same amount of energy.

On the good news side, I urge you to look into the current PR #99 :-) Similar speedup for Adadelta *and* better convergence ...
